texar/modules/encoders/xlnet_encoder.py

```diff
@@ -0,0 +1,480 @@
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from texar.core import layers
+from texar.hyperparams import HParams
+from texar.modules.pretrained import (xlnet_utils, XLNetBase, PositionWiseFF,
+                                      RelativePositionalEncoding,
+                                      RelativeMultiheadAttention)
+from texar.utils import dict_fetch
+
+
+__all__ = [
+    "XLNetEncoder",
+]
+
+
+class XLNetEncoder(XLNetBase):
+    r"""Raw XLNet module for encoding sequences.
+
+    This module supports the architecture first proposed
+    in `(Yang et al.)` XLNet.
+
+    Args:
+        pretrained_model_name (optional): a str with the name
+            of a pre-trained model to load selected in the list of:
+            `xlnet-base-cased`, `xlnet-large-cased`.
+            If `None`, will use the model name in :attr:`hparams`.
+        cache_dir (optional): the path to a folder in which the
+            pre-trained models will be cached. If `None` (default),
+            a default directory will be used.
+        hparams (dict or HParams, optional): Hyperparameters. Missing
+            hyperparameter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter structure
+            and default values.
+        init (optional): whether to initialize `XLNetEncoder`.
+    """
+
+    def __init__(self,
+                 pretrained_model_name: Optional[str] = None,
+                 cache_dir: Optional[str] = None,
+                 hparams=None,
+                 init=True):
+
+        super().__init__(pretrained_model_name=pretrained_model_name,
+                         cache_dir=cache_dir,
+                         hparams=hparams)
+
+        if self.pretrained_model_dir:
+            self._hparams = HParams(self.pretrained_model_hparams,
+                                    self._hparams.todict())
+
+        num_layers = self._hparams.num_layers
+        num_heads = self._hparams.num_heads
+        head_dim = self._hparams.head_dim
+
+        self.word_embed = nn.Embedding(self._hparams.vocab_size,
+                                       self._hparams.hidden_dim)
+        self.pos_embed = RelativePositionalEncoding(
+            hparams={
+                "dim": self._hparams.hidden_dim,
+                "max_seq_len": self._hparams.max_seq_len,
+            })
+        self.dropout = nn.Dropout(self._hparams.dropout)
+
+        self.r_r_bias: Optional[nn.Parameter]
+        self.r_w_bias: Optional[nn.Parameter]
+        self.r_s_bias: Optional[nn.Parameter]
+
+        if not self._hparams.untie_r:
+            self.r_r_bias = nn.Parameter(torch.Tensor(num_heads, head_dim))
+            self.r_w_bias = nn.Parameter(torch.Tensor(num_heads, head_dim))
+            self.r_s_bias = (nn.Parameter(torch.Tensor(num_heads, head_dim))
+                             if self._hparams.use_segments else None)
+        else:
+            self.r_r_bias = None
```

Comment:
Something I just learnt last week: if some variable should have type `Optional[T]`, instead of writing it like this, we can just do:
```python
var = None
if ...:
    var = something_else
```
This is much more concise and can also pass type-check.